### ISSUE TITLE
New rule proposal: disallow === -0

### ISSUE THREAD

**Issue by domenic** [eslint/eslint#8066]
**Please describe what the rule should do:**

The rule should warn against code that tries to use the `===` operator to compare against `-0`, since that will not work as intended. That is, code like `x === -0` will pass for both `+0` and `-0`. The author probably intended `Object.is(x, -0)`.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// Don't accept negative zero
if (x === -0) {
  throw new TypeError("-0 is not a valid value for this API");
}

// This code intends to only throw on -0, but instead throws on +0 as well.
```

```js
// Allow any negative number, including negative zero
if (x === -0 || x < 0) {
  // use x
}

// This code intends to disallow +0, but instead allows it through.
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This rule seems like a fairly universal potential mistake. It meets all the core rule guidelines and is a perfect candidate for linting.

**Comment by ilyavolodin** [comment]
I think this an incredibly rare situation. I've honestly never seen a comparison like that. Not sure if this fits the core.

**Comment by domenic** [comment]
Hmm. Here is the bug in the wild that inspired me to suggest this: https://github.com/jsdom/webidl-conversions/blob/6e291bbaffc15c0bd3e311d51edbf2973c46a915/lib/index.js#L93

To me this is very similar to the use-isnan rule.

**Comment by platinumazure** [comment]
I tend to agree with @ilyavolodin but could be persuaded otherwise.

@domenic If we don't accept this proposal for ESLint core, please note that you are more than welcome to create a custom rule or plugin and so be able to lint for your desired use case. I think it would be a great idea to find or create a plugin about comparisons that will lead to unexpected results, such as this case.

**Comment by nzakas** [comment]
I can see an argument for this as a core rule, as it's one of those language quirks that you could spend days trying to debug if you misunderstood what happens, and I agree with the comparison to use-isnan, as that rule solves a similar problem. 

I'd even go so far as to say I think it should be included in "eslint:recommended". 

I'll champion. 

@domenic are you willing to submit a PR for this rule if the team accepts it?

**Comment by domenic** [comment]
After looking through some recent PRs for new rules, that seems doable; I'd be happy to contribute the implementation/docs.

**Comment by ljharb** [comment]
@domenic What's the bug in the code you linked? Whether `x` is positive or negative zero, it will still return a positive zero, which the comment indicates was the desired effect.

**Comment by domenic** [comment]
@ljharb you're right that in this case it happened to work, but that's only because there are not any further steps in that algorithm after the `if()` block. If further steps were added (as they were in a subsequent refactoring I was in the process of making while I noticed the bug), then the code would return too early, whereas in fact it was supposed to proceed onward to the next set of steps.

**Comment by ljharb** [comment]
Makes sense, thanks for explaining. Looking forward to enabling this rule!

**Comment by nzakas** [comment]
@eslint/eslint-team can we get a couple more :+1:s for this so we can move forward?

**Comment by vitorbal** [comment]
Added my support.

**Comment by nzakas** [comment]
@domenic looks like you are good to go. If you have never created a rule before, I'd suggest using the [Yeoman generator](https://github.com/eslint/generator-eslint) to get started (alternately, you can copy an existing rule's files and modify them). 

We can't add this rule into "ESLint:recommended" without a major release, so don't worry about that as we will handle it when we release 4.0.0.

**Comment by domenic** [comment]
Sounds great, thanks! I'll most likely work on this over the upcoming weekend.

**Comment by vitorbal** [comment]
Thanks, @domenic ! If you have any questions, you can always pop over to our [Gitter chat room](https://gitter.im/eslint/eslint) and we'll be happy to help!

**Comment by Aladdin-ADD** [comment]
@domenic I've done some work about this rule. Since I'm a new developer to this project, could you  give some advice?

**Comment by ljharb** [comment]
Looks like this can be closed, since #8091 is merged?

**Comment by not-an-aardvark** [comment]
Yep
